### PR TITLE
RHSSO Helper Refactor

### PR DESCRIPTION
**Pull request by ColeHiggins2**
Draft work for RHSSO Helper Refactor

### CODE REVIEW

File: robottelo/utils/sso.py
Comment (mshriver) [inline robottelo/utils/sso.py:20]: docblock here please to explain why you're injecting settings.rhsso and what you're expecting to read from it.

File: robottelo/utils/sso.py
Comment (mshriver) [inline robottelo/utils/sso.py:41]: f-strings throughout please

**Review by JacobCallahan — CHANGES_REQUESTED** [review]
Just some small suggestions. Really great addition!

**Comment by latran** [comment]
@ColeHiggins2 Any updates on this from the changes suggested?

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

File: robottelo/utils/sso.py
Comment (mshriver) [inline robottelo/utils/sso.py:19]: @JacobCallahan I'm starting to think `robottelo.hosts` should be a package, not a module.

File: robottelo/utils/sso.py
Comment (mshriver) [inline robottelo/utils/sso.py:197]: This instance needs to be provided by a fixture, not a module defined globally importable one.

File: robottelo/utils/sso.py
Comment (mshriver) [inline robottelo/utils/sso.py:190]: This should reference `self.satellite` - but requires basing the class off `ContentHost` to have access.

otherwise, any functions dealing with Satellite need to take a required `sat` kwarg.

File: tests/foreman/destructive/test_ldap_authentication.py
Comment (mshriver) [inline tests/foreman/destructive/test_ldap_authentication.py:39]: When this instance is provided via a fixture, the target sat (at the relevant scope) should be passed in.

```
@pytest.fixture
def sso_host(target_sat):
    return SSOHost(satellite=target_sat)
```

File: robottelo/utils/sso.py
Comment (mshriver) [inline robottelo/utils/sso.py:87]: This pattern is copied from the existing module, and should not be continued if you're here to improve the helpers state.

This constant is imported from the module, and then mutated - constants should be immutable (at least treated as such).

Consider creating a basic class, or even just a namedtuple, as an instance attribute. Then it's just tied to this one host, and can be manipulated however is necessary without impacting anything else running in parallel.

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

File: robottelo/utils/sso.py
Comment (JacobCallahan) [reply robottelo/utils/sso.py:19]: oooff that's something we can think about later...

**Comment by omkarkhatavkar** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_single_sign_on_using_rhsso

**Comment by omkarkhatavkar** [comment]
@ColeHiggins2 can you please rebase i was able to fix many tests with rhsso

File: robottelo/utils/sso.py
Comment (omkarkhatavkar) [inline robottelo/utils/sso.py:147]: we can use @property all of those getter methods and with the use of self.hostname

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso

**Comment by ColeHiggins2** [comment]
Opening back up for review. tests will fail until https://github.com/SatelliteQE/robottelo/pull/10033 is merged

**Comment by ColeHiggins2** [comment]
Results with https://github.com/SatelliteQE/robottelo/pull/10033 added in

Launching pytest with arguments /home/colehiggins/projects/robottelo/tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso in /home/colehiggins/projects/robottelo

============================= test session starts ==============================

collecting ... collected 1 item

tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso 

=============================== warnings summary ===============================
../robo38/lib64/python3.8/site-packages/_pytest/config/__init__.py:1252
  /home/colehiggins/projects/robo38/lib64/python3.8/site-packages/_pytest/config/__init__.py:1252: PytestConfigWarning: Unknown config option: rp_ignore_errors
  
    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================== 1 passed, 1 warning in 1012.34s (0:16:52) ===================

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso

File: robottelo/utils/sso.py
Comment (jyejare) [inline robottelo/utils/sso.py:32]: According to our helpers refactor history and best practices, , the utils module is for external tools that helps pytest/robottelo to execute the tests or which are not related to host objects or not satellite endpoint objects etc. 

So If the helpers are associated with host objects as we are having here then the best place for them is `robottelo/hosts.py`. Unfortunately `robottelo/hosts.py` is already a big file containing Chost, Sat and Capsule classes and its methodsssssss! 

So we should just initialise the SSOHost in that file and you should move all SSOHost methods to `SSOHost_mixins.py` categorised in classes and then inherit those classes/methods in SSOHost class. Refer https://github.com/SatelliteQE/robottelo/blob/master/robottelo/hosts.py#L123 where ContentHost class importing ContentHost mixins class. Also I agree with @mshriver comment of inheriting ContentHost class than Host class.

@mshriver @JacobCallahan  We should move all methods of CHost, Sat and Capsule classes methods to the mixins and just initialise classes in hosts.py. Its a separate effort and not blocking this PR for it!

File: robottelo/utils/sso.py
Comment (jyejare) [inline robottelo/utils/sso.py:41]: Use f string ?

File: robottelo/utils/sso.py
Comment (jyejare) [inline robottelo/utils/sso.py:91]: f-string !

File: tests/foreman/destructive/test_ldap_authentication.py
Comment (jyejare) [inline tests/foreman/destructive/test_ldap_authentication.py:322]: Looks like accidentally you pushed conflicted code. Please rectify.

**Review by jyejare — CHANGES_REQUESTED** [review]
@ColeHiggins2 Overall all looks good with few advices.

1. Use `robottelo/hosts.py` for initialising SSOHost class. Check inline comment for more details.
2. Remove `robottelo/rhsso_utils.py` module in this PR itself.
3. Use f-strings
4. Fix conflict strings appeared in your code.

Optional:
1. Looks like we are using so much of SSO CLI command to execute, should we put efforts in using `python SSO API client` instead. Not blocking this PR for it but a future advice for component owner !

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso

File: robottelo/utils/sso.py
Comment (omkarkhatavkar) [reply robottelo/utils/sso.py:32]: `class SSOHost` without 'Satellite' class has no value having the composition of Satellite class with SSOHost class make sense and not need to pass the satellite object in every function. As @mshriver said it is good to have self.satellite attribute associated with rhsso host class

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso

File: robottelo/utils/sso.py
Comment (omkarkhatavkar) [inline robottelo/utils/sso.py:170]: change this to `get_two_factor_token_rh_sso_url `

**Comment by ColeHiggins2** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_ldap_authentication.py::test_external_new_user_login_and_check_count_rhsso

**Comment by jyejare** [comment]
@ColeHiggins2 Please resolve conflicts !

File: robottelo/utils/sso.py
Comment (ColeHiggins2) [reply robottelo/utils/sso.py]: Fixes made. After speaking with jake, We decided to keep all contents of ssohost() in hosts.py and not break it up into a mixins file.

File: robottelo/hosts.py
Comment (jyejare) [inline robottelo/hosts.py:1768]: Add a docstring got this host type !

File: robottelo/hosts.py
Comment (jyejare) [inline robottelo/hosts.py:1775]: I dont see this variable is used at all, please remove it !

Comment (jyejare) [reply robottelo/hosts.py:1775]: Also sat_obect is also not used anywhere. TBH I am not sure why we need sat_obj as a parameter. Isnt the SSOHost inheriting the Satellite class could be used as `self.hostname` to identify that satellite object.

File: robottelo/hosts.py
Comment (jyejare) [inline robottelo/hosts.py:1931]: You dont need to initialize SSOHost here, it should be initialized while in use because one can pass the hostname to SSOHost.

File: robottelo/hosts.py
Comment (jyejare) [inline robottelo/hosts.py:1929]: TBH I really didnt like how many helpers we have added this in one file, instead this should have been externalized to in Mixins to track them separately.

@JacobCallahan !

**Review by jyejare — CHANGES_REQUESTED** [review]
ACK pending the comments !

File: robottelo/hosts.py
Comment (omkarkhatavkar) [reply robottelo/hosts.py:1775]: I personally think SSOHost should not inherit the Satellite Class, It would rather have the Satellite object as part of the composition rather than inheriting.

File: robottelo/hosts.py
Comment (omkarkhatavkar) [reply robottelo/hosts.py:1931]: Use above line in the fixture as Mike mentioned earlier

File: robottelo/hosts.py
Comment (JacobCallahan) [reply robottelo/hosts.py:1929]: I believe it was previously requested for @ColeHiggins2 to move this class into hosts.py
With that in mind, I don't think there is enough in the class definition to justify externalizing to a mixin.

**Comment by ColeHiggins2** [comment]
Work done in the last push:

-Removed sso_host = SSOHost() from robottelo/hosts.py
-updated default_sso_host fixture in Satellite_auth.py 
-Removed sat_obj as an argument and replaced all uses with self.satellite
-removed module_target_sat from any default_sso_host.function() call in the corresponding test files

**Comment by jyejare** [comment]
@ColeHiggins2 You need to resolve the conflicts !

**Review by jyejare — APPROVED** [review]
ACK with instructions to rebase the code before merge !

**Comment by jyejare** [comment]
@ColeHiggins2 Please rebase the code and resolve the conflicts, before we merge the code !

**Comment by jyejare** [comment]
@ColeHiggins2 Please rebase the PR as we are going to merge it !

**Comment by ColeHiggins2** [comment]
Launching pytest with arguments /home/colehiggins/projects/robottelo/tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist in /home/colehiggins/projects/robottelo

============================= test session starts ==============================
platform linux -- Python 3.8.2, pytest-7.1.3, pluggy-0.13.1 -- /home/colehiggins/projects/robo38/bin/python3.8
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/colehiggins/projects/robottelo, configfile: pyproject.toml
plugins: reportportal-5.1.2, xdist-2.5.0, ibutsu-2.2.4, forked-1.3.0, mock-3.10.0, services-2.2.1
collecting ... collected 1 item

tests/foreman/destructive/test_ldap_authentication.py::test_login_failure_rhsso_user_if_internal_user_exist 

=============================== warnings summary ===============================
.
.
.

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================== 1 passed, 2 warnings in 681.16s (0:11:21) ===================

**Review by JacobCallahan — APPROVED** [review]
ACK, great work!!

File: tests/foreman/destructive/test_ldap_authentication.py
Comment (omkarkhatavkar) [inline tests/foreman/destructive/test_ldap_authentication.py:411]: just remove this `default_sso_host.get_rhsso_client_id()` as not required here, and move that as first line to get the latest token

**Review by omkarkhatavkar — APPROVED** [review]
ACK, pending one comment to resolve

**Pull request merged**
